langchain-core/src/runnables/graph_mermaid.ts

```diff
@@ -97,22 +97,38 @@ export function drawMermaid(
 
   const seenSubgraphs = new Set<string>();
 
+  // sort prefixes by path length for correct nesting
+  function sortPrefixesByDepth(prefixes: string[]): string[] {
+    return [...prefixes].sort((a, b) => {
+      return a.split(":").length - b.split(":").length;
+    });
+  }
+
   function addSubgraph(edges: Edge[], prefix: string): void {
     const selfLoop = edges.length === 1 && edges[0].source === edges[0].target;
     if (prefix && !selfLoop) {
       const subgraph = prefix.split(":").pop()!;
-      if (seenSubgraphs.has(subgraph)) {
-        throw new Error(
-          `Found duplicate subgraph '${subgraph}' -- this likely means that ` +
-            "you're reusing a subgraph node with the same name. " +
-            "Please adjust your graph to have subgraph nodes with unique names."
-        );
+
+      if (seenSubgraphs.has(prefix)) {
```

Comment:
We should probably still throw in this case?

Reply:
Good point; I restored this, and also needed to add one more protection to the recursion to prevent reaching deep paths twice